### PR TITLE
Add a /ignore command

### LINKED ISSUE #80: Ignore Command
I'd like to see a /ignore command, just like Essentials.
I'm sure you are already planning on this, but just in case!

### PR DESCRIPTION
**Pull request by Jaimss**
Resolves #80

### Commands
Added `/ignore <player>`

### Permissions
Permission: `chatchat.ignore`
Bypass Permission: `chatchat.ignore.bypass`

#### Potential Changes: 
- [ ] Ignored players can't see your chat. Currently, it only blocks their chat from getting to you. This would block chat both ways
- [ ] Change the way `<player>` works in the command's success message. Currently just uses Player#getDisplayName(). Could instead allow PAPI Placeholders so they could use like %vault_displayname% for example.

#### Notes
This probably shouldn't be merged until the storage system is setup that way ignored players will be saved across restarts. Marked as draft for now.

### CODE REVIEW

**Comment by Jaimss** [comment]
I need filter ignored users in the Channel#targets

File: api/src/main/java/at/helpch/chatchat/api/ChatUser.java [hunk 1]
```diff
@@ -5,6 +5,8 @@
 import org.jetbrains.annotations.Nullable;
 
 import java.util.Optional;
+import java.util.Set;
```
Comment (BlitzOffline) [inline api/src/main/java/at/helpch/chatchat/api/ChatUser.java:8]: Forgot to optimize imports?

File: plugin/src/main/java/at/helpch/chatchat/user/ChatUserImpl.java [hunk 2]
```diff
@@ -77,8 +81,6 @@ public boolean privateMessages() {
     public void privateMessages(final boolean enabled) {
         this.privateMessages = enabled;
     }
-
-    @Override
```
Comment (BlitzOffline) [inline plugin/src/main/java/at/helpch/chatchat/user/ChatUserImpl.java:81]: why remove the override?

File: plugin/src/main/java/at/helpch/chatchat/channel/ChatChannel.java [hunk 3]
```diff
@@ -41,25 +42,36 @@ public static void defaultChannel(@NotNull final ChatChannel toSet) {
...
     }
 
     @Override
     public Set<User> targets(final @NotNull User source) {
+
+        final Predicate<User> filterIgnores = user -> {
+            // console can't be ignored
+            if (source instanceof ConsoleUser) return true;
+            return !user.ignoredUsers().contains(source.uuid()) ||
+                ((ChatUser) source).player().hasPermission(IgnoreCommand.IGNORE_BYPASS_PERMISSION);
```
Comment (BlitzOffline) [inline plugin/src/main/java/at/helpch/chatchat/channel/ChatChannel.java:60]: We might want to split these 2 and add an extra check to see if the user is actually a ChatUser. People could add their own users (in the future :) ) so we might end up with some exceptions thrown here.

**Review by BlitzOffline — CHANGES_REQUESTED** [review]
Looks good just a few small changes

File: plugin/src/main/java/at/helpch/chatchat/user/ChatUserImpl.java (hunk 2 above)
Comment (Jaimss) [reply plugin/src/main/java/at/helpch/chatchat/user/ChatUserImpl.java:81]: whoops, I think i did that when fixing merge conflicts probably

File: plugin/src/main/java/at/helpch/chatchat/channel/ChatChannel.java (hunk 3 above)
Comment (Jaimss) [reply plugin/src/main/java/at/helpch/chatchat/channel/ChatChannel.java]: Is the recent commit better? I think that's what you're asking for

**Review by BlitzOffline — APPROVED** [review]

File: plugin/src/main/java/at/helpch/chatchat/command/IgnoreCommand.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+import at.helpch.chatchat.api.ChatUser;
+import dev.triumphteam.cmd.bukkit.annotation.Permission;
+import dev.triumphteam.cmd.core.BaseCommand;
+import dev.triumphteam.cmd.core.annotation.Command;
+import dev.triumphteam.cmd.core.annotation.Default;
+
+@Command("ignore")
+public class IgnoreCommand extends BaseCommand {
+
+    final ChatChatPlugin plugin;
```
Comment (Kqliber) [inline plugin/src/main/java/at/helpch/chatchat/command/IgnoreCommand.java:13]: private pls

File: plugin/src/main/java/at/helpch/chatchat/command/IgnoreCommand.java [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    @Permission(IGNORE_PERMISSION)
+    @Default
+    public void ignore(ChatUser sender, ChatUser target) {
+        final var ignored = sender.ignoredUsers().contains(target.uuid());
+
+        final var messageHolder = plugin.configManager().messages();
+        final var message = ignored ? messageHolder.unignoredPlayer() : messageHolder.ignoredPlayer();
+
+        if (ignored) sender.unignoreUser(target);
+        else sender.ignoreUser(target);
```
Comment (Kqliber) [inline plugin/src/main/java/at/helpch/chatchat/command/IgnoreCommand.java:30]: add braces

File: plugin/src/main/java/at/helpch/chatchat/data/impl/gson/ChatUserAdapter.java [hunk 6]
```diff
@@ -30,6 +34,10 @@ public void write(JsonWriter out, ChatUser value) throws IOException {
         out.value(value.privateMessages());
         out.name("social-spy");
         out.value(value.socialSpy());
+        out.name("ignored-users");
+        out.beginArray();
+        for (UUID uuid : value.ignoredUsers()) out.value(uuid.toString());
```
Comment (Kqliber) [inline plugin/src/main/java/at/helpch/chatchat/data/impl/gson/ChatUserAdapter.java:39]: and here too

**Review by Kqliber — COMMENTED** [review]
fix merge conflicts too and ill merge

**Pull request merged**